tool/src/kotlin/mod.rs

```diff
@@ -0,0 +1,1265 @@
+            Type::Struct(ref strct) => {
+                let op_id = strct.id();
+                self.formatter.fmt_type_name(op_id)
+            }
+            Type::Enum(ref enum_def) => self.formatter.fmt_type_name(enum_def.tcx_id.into()),
+            Type::Slice(hir::Slice::Str(_, _)) => self.formatter.fmt_string().into(),
+            Type::Slice(hir::Slice::Primitive(_, ty)) => {
+                self.formatter.fmt_primitive_slice(ty).into()
+            }
+
+            Type::Slice(hir::Slice::Strs(_)) => self.formatter.fmt_str_slices().into(),
+            _ => unreachable!("unknown AST/HIR variant"),
+        }
+    }
+}
+
+type MethodLtMap<'a> = BTreeMap<Lifetime, BorrowedLifetimeInfo<'a>>;
+
+#[derive(Template)]
+#[template(path = "kotlin/Method.kt.jinja", escape = "none")]
+struct MethodTpl<'a> {
+    // todo: comment: String,
+    declaration: String,
+    /// The C method name
+    native_method_name: Cow<'a, str>,
+
+    /// Conversion code for each parameter
+    param_conversions: Vec<Cow<'a, str>>,
+    return_expression: Option<Cow<'a, str>>,
+    writeable_return: bool,
+    slice_conversions: Vec<Cow<'a, str>>,
+}
+
+struct NativeMethodInfo {
+    declaration: String,
+}
+
+#[cfg(test)]
+mod test {
+
+    use diplomat_core::hir::TypeDef;
+    use quote::quote;
+
+    use crate::common::ErrorStore;
+
+    use super::formatter::test::new_tcx;
+    use super::{formatter::KotlinFormatter, TyGenContext};
+
+    #[test]
+    fn test_enum() {
+        let tk_stream = quote! {
+            #[diplomat::bridge]
+            mod ffi {
+
+                pub enum Cont {
+                    A,
+                    B,
+                    C,
+                    D,
+                }
+
+                pub enum ContNumbered {
+                    Alpha=0,
+                    Beta=1,
+                    Gamma=2,
+                }
+
+                pub enum NonCont {
+                    Aleph=0,
+                    Bet=1,
+                    Tav=22,
+                }
+
+                pub enum Neg {
+                    Neg3=-3,
+                    Neg1=-1,
+                    Thirteen=13,
+                }
+
+            }
+        };
+
+        let tcx = new_tcx(tk_stream);
+        let mut all_types = tcx.all_types();
+        while let Some((type_id, TypeDef::Enum(enum_def))) = all_types.next() {
+            let error_store = ErrorStore::default();
+            let formatter = KotlinFormatter::new(&tcx, None);
+            let mut ty_gen_cx = TyGenContext {
+                tcx: &tcx,
+                formatter: &formatter,
+                errors: &error_store,
+            };
+            let type_name = enum_def.name.to_string();
+            // test that we can render and that it doesn't panic
+            let (_, boop) =
+                ty_gen_cx.gen_enum_def(enum_def, type_id, &type_name, "dev.gigapixel", "somelib");
+            println!("{boop}")
+        }
+    }
+
+    #[test]
+    fn test_struct() {
```

Comment:
suggestion: you can use insta tests for these things. You can see how tool/src/js does it with the snapshots, and diplomat_core uses insta a lot.

Reply:
It might take me a while to get to know insta. But I'll have a look.